### PR TITLE
Json Stringify option to not write out null optional fields

### PR DESCRIPTION
**Pull request by hazeycode**
Fixed version of https://github.com/ziglang/zig/pull/8908 with added test to parse stringify output

### CODE REVIEW

**Comment by hazeycode** [comment]
Found another failure case. Needs more testing.

**Comment by daurnimator** [comment]
I wonder if a better option might be a `skip_defaults` to avoid rendering all defaults.
Then you could opt-in per optional field by setting their default value to `null`.

File: lib/std/json.zig [hunk 1]
```diff
@@ -2731,6 +2731,9 @@ pub const StringifyOptions = struct {
 
             /// Should unicode characters be escaped in strings?
             escape_unicode: bool = false,
+
+            /// Should optional fields with null value be written?
+            write_null_optional_fields: bool = true,
```
Comment (daurnimator) [inline lib/std/json.zig:2736]: Why is this under `StringOutputOptions`?

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: lib/std/json.zig (hunk 1 above)
Comment (hazeycode) [reply lib/std/json.zig:2736]: Good question. This was an error on my part

**Comment by hazeycode** [comment]
> I wonder if a better option might be a `skip_defaults` to avoid rendering all defaults.
> Then you could opt-in per optional field by setting their default value to `null`.

I'm unconvinced. It does make sense but is there a use-case where you don't want to stringify nulls that go OTW but would want to stringify some local default that doesn't make sense remotely? i.e.

```
const Message = struct {
    user_id: u32 = 1337, // should stringify
    optional: []const u8 = null, // should not stringify 
    required: []const u8, // should stringify
}
```

Of course, the problem here is easily avoided, however, not convinced that it's a good user experience.

**Comment by hazeycode** [comment]
This appears to be working properly now!
Commits likely want squashing but I had all kinds of bother trying to rebase

File: lib/std/json.zig [hunk 2]
```diff
@@ -2822,23 +2825,36 @@ pub fn stringify(
...
 
-                if (!field_output) {
-                    field_output = true;
-                } else {
-                    try out_stream.writeByte(',');
-                }
-                if (child_options.whitespace) |child_whitespace| {
-                    try out_stream.writeByte('\n');
-                    try child_whitespace.outputIndent(out_stream);
+                var emit_field = true;
```
Comment (andrewrk) [inline lib/std/json.zig:2828]: Looks redundant with `field_output`

**Review by andrewrk — APPROVED** [review]

**Pull request merged**